File: Data Structures & Algorithms/target-sum/submission-0.cpp
class Solution {
public:
    int ans = 0;

    void backtrack(int i, int sum, vector<int>& nums, int target)
    {
        if (i == nums.size())
        {
            if (sum == target)
                ans++;
            return;
        }

        backtrack(i + 1, sum + nums[i], nums, target);
        backtrack(i + 1, sum - nums[i], nums, target);
    }

    int findTargetSumWays(vector<int>& nums, int target)
    {
        backtrack(0, 0, nums, target);
        return ans;
    }
};

File: Data Structures & Algorithms/combination-target-sum/submission-0.py
class Solution:
    def combinationSum(self, nums: List[int], target: int) -> List[List[int]]:
        if len(nums)==0:
            return ans.append([])
        ans=[]
        subset=[]
        def backtrack(i,subset):
            if sum(subset)==target and subset not in ans:
                ans.append(subset.copy())
                return
            if i==len(nums) or sum(subset)>target:
                return

            subset.append(nums[i])
            backtrack(i,subset)
            subset.pop()
            backtrack(i+1,subset)

        backtrack(0,subset)
        return ans


        